### DISCUSSION TITLE
driver.Rmd to schedule pipeline beside the app

### DISCUSSION THREAD

**Discussion by SaJaToGu** [wlandau/targets-shiny#11] [Ideas]
Dear Mr. Landau,
would it be possible to have a driver.Rmd or pipeline.Rmd beside the app.R to schedule the pipeline on e.g. the RStudio Connect server and use the App on the server in addition for observation?
I tried to use
source("R/project.R")
source("R/process.R")
process_run()
but run into the shiny problem.
Error: shinyjs: could not find the Shiny session object.

May be your example could be setup in a way that it checks if within a Shiny session or not to avoid shiny commands?

Best regards,
Guido

**Comment by wlandau** [comment]
This app is designed to configure and deploy the pipeline in a fully interactive way rather than on a pre-committed schedule. For a schedule, I would suggest a different approach like setting up an ordinary `targets` pipeline that is invoked from a simple Rmd file like you mentioned. For that use case, a Shiny app is not needed. https://solutions.rstudio.com/r/workflows/ describes a successful pattern. If both scheduling and the Shiny app are both required, then a more complicated approach is required. You could deploy both the app and the driver script separately and have them interact with the same mounted data store, for example. But because it is nontrivial, I prefer to keep this particular example Shiny app separate from those concerns.

**Reply by SaJaToGu** [reply]
Dear Mr. Landau,
may thanks for your reply.
I thought about the more complicated approach because it's easier to identify error with your stdin/stdout approch and the simple Rmd file deployed to RStudio Connect is not stored in the persistent folder and also not the meta data.
 That I like with your app approach. I also thought maybe to use the graph to select the objects and show their table inside the app.
 The code in my first message did run but the following worked.
 But I do not know why it's behavior is different.
`
```{r run pipeline}
if (!project_exists()){
  project_create("covForecastAccuracy")
  project_save("covForecastAccuracy")
}
#write_pipeline("covForecastAccuracy")
    args <- list(
      # Important! Garbage collection should not terminate the process:
      cleanup = FALSE,
      # supervise = transient_active(), # Otherwise, the process quits on logout.
      stdout = project_stdout(),
      stderr = project_stderr()
    )
    script <- project_path(project_get(), "_targets.R")
    store <- project_path(project_get(), "_targets")
    # Here is where we actually run the pipeline:
    # multicore possible for RStudio
    Sys.setenv(R_PARALLELLY_FORK_ENABLE = TRUE)
    px <- tar_make_future(
      workers = 10,
      callr_function = r_bg,
      callr_arguments = args,
      script = script,
      store = store
    )
    # Do not give back control until the pipeline write a _targets/meta/process
    # file with the PID of the main process.
    while (process_not_done(px) && !tar_exist_process(store)) Sys.sleep(0.05)
    # Do not give back control until the PID in _targets/meta/process
    # agrees with the PID of the process handle we have in memory.
    while (process_not_done(px) && !process_agrees(px)) Sys.sleep(0.05)

```
`
